test/content-routing/content-routing.node.js

```diff
@@ -0,0 +1,327 @@
+'use strict'
+/* eslint-env mocha */
+
+const chai = require('chai')
+chai.use(require('dirty-chai'))
+const { expect } = chai
+const nock = require('nock')
+const sinon = require('sinon')
+
+const pDefer = require('p-defer')
+const mergeOptions = require('merge-options')
+
+const CID = require('cids')
+const DelegatedContentRouter = require('libp2p-delegated-content-routing')
+const multiaddr = require('multiaddr')
+
+const peerUtils = require('../utils/creators/peer')
+const { baseOptions, routingOptions } = require('./utils')
+
+describe('content-routing', () => {
+  describe('no routers', () => {
+    let node
+
+    before(async () => {
+      [node] = await peerUtils.createPeer({
+        config: baseOptions
+      })
+    })
+
+    it('.findProviders should return an error', async () => {
+      try {
+        for await (const _ of node.contentRouting.findProviders('a cid')) {} // eslint-disable-line no-unused-vars
+        throw new Error('.findProviders should return an error')
+      } catch (err) {
+        expect(err).to.exist()
+        expect(err.code).to.equal('NO_ROUTERS_AVAILABLE')
+      }
+    })
+
+    it('.provide should return an error', async () => {
+      await expect(node.contentRouting.provide('a cid'))
+        .to.eventually.be.rejected()
+        .and.to.have.property('code', 'NO_ROUTERS_AVAILABLE')
+    })
+  })
+
+  describe('via dht router', () => {
+    const number = 5
+    let nodes
+
+    before(async () => {
+      nodes = await peerUtils.createPeer({
+        number,
+        config: routingOptions
+      })
+
+      // Ring dial
+      await Promise.all(
+        nodes.map((peer, i) => peer.dial(nodes[(i + 1) % number].peerInfo))
+      )
+    })
+
+    afterEach(() => {
+      sinon.restore()
+    })
+
+    after(() => Promise.all(nodes.map((n) => n.stop())))
+
+    it('should use the nodes dht to provide', () => {
+      const deferred = pDefer()
+
+      sinon.stub(nodes[0]._dht._dht, 'provide').callsFake(() => {
+        deferred.resolve()
+      })
+
+      nodes[0].contentRouting.provide()
+      return deferred.promise
+    })
+
+    it('should use the nodes dht to find providers', async () => {
+      const deferred = pDefer()
+
+      sinon.stub(nodes[0]._dht._dht, 'findProviders').callsFake(function * () {
+        deferred.resolve()
+        yield
+      })
+
+      for await (const _ of nodes[0].contentRouting.findProviders()) { // eslint-disable-line no-unused-vars
+        break
+      }
```

Comment:
Is this for loop necessary? It waits for the first call and then exists the for loop, but we really only care about waiting for the deferred. Shouldn't you just be able to have `nodes[0].contentRouting.findProviders()` with no await?

Reply:
I can await for next. Just call the `findProviders` does not iterate